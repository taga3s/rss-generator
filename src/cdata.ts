/**
 * Encloses the given data in a CDATA section.
 */
export const cdata = (data: string): string => {
  if (data.includes("]]>")) {
    throw new Error('CDATA section contains forbidden string "]]>"');
  }
  return `<![CDATA[${data}]]>`;
};
