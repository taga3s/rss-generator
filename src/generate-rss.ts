import { stringify } from "./stringify.ts";
import { createXMLTree } from "./ast.ts";
import type { Channel, Item, Namespaces } from "./generate-rss_types.ts";
import { buildXMLObj } from "./xmlobj.ts";

/**
 * Generate RSS feed
 * @param data - RSS feed data
 * @returns RSS feed as string
 */
export const generateRSS = (
  data: { channel: Channel; items?: Item[]; namespaces?: Namespaces },
): string => {
  const { channel, items, namespaces } = data;

  const xmlObj = buildXMLObj({ channel, items, namespaces });

  const xmlTree = createXMLTree(xmlObj);

  return stringify(xmlTree);
};
