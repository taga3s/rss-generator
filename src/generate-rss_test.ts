import { cdata } from "./cdata.ts";
import { generateRSS } from "./generate-rss.ts";
import type { Channel, Item, Namespaces } from "./generate-rss_types.ts";
import { assertSnapshot } from "jsr:@std/testing/snapshot";

const fixture = (): {
  xml: string;
} => {
  const channel: Channel = {
    title: "Example Web",
    link: "https://example.com",
    description: cdata("Example description"),
    ttl: 60,
    language: "en",
    category: ["sports", "politics", "technology"],
    copyright: "Example Web",
  };

  const items: Item[] = [
    {
      title: "Example Title 1",
      description: cdata("Example description"),
      content: {
        encoded: cdata("<p>Example content</p>"),
      },
      link: "https://example.com/articles/1",
      guid: {
        isPermaLink: true,
        value: "https://example.com/articles/1",
      },
      pubDate: "Sun, 27 Apr 2025 03:24:20 GMT",
    },
    {
      title: "Example Title 2",
      description: cdata("Example description"),
      content: {
        encoded: cdata("<p>Example content</p>"),
      },
      link: "https://example.com/articles/2",
      guid: {
        isPermaLink: true,
        value: "https://example.com/articles/2",
      },
      pubDate: "Sun, 27 Apr 2025 03:24:20 GMT",
    },
  ];

  const namespaces: Namespaces = ["atom", "content", "dc", "slash"];

  return {
    xml: generateRSS({ channel, items, namespaces }),
  };
};

Deno.test("generateRSS", async (t) => {
  const { xml } = fixture();
  await assertSnapshot(t, xml);
});
