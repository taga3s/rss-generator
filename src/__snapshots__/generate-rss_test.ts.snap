export const snapshot = {};

snapshot[`generateRSS 1`] = `
'<?xml version="1.0" encoding="UTF-8"?>
<rss version="2.0" xmlns:atom="http://www.w3.org/2005/Atom" xmlns:content="http://purl.org/rss/1.0/modules/content/" xmlns:dc="http://purl.org/dc/elements/1.1/" xmlns:slash="http://purl.org/rss/1.0/modules/slash/">
  <channel>
    <title>Example Web</title>
    <description><![CDATA[Example description]]></description>
    <link>https://example.com</link>
    <category>sports</category>
    <category>politics</category>
    <category>technology</category>
    <copyright>Example Web</copyright>
    <language>en</language>
    <ttl>60</ttl>
    <item>
      <title>Example Title 1</title>
      <description><![CDATA[Example description]]></description>
      <content:encoded><![CDATA[<p>Example content</p>]]></content:encoded>
      <link>https://example.com/articles/1</link>
      <guid isPermaLink="true">https://example.com/articles/1</guid>
      <pubDate>Sun, 27 Apr 2025 03:24:20 GMT</pubDate>
    </item>
    <item>
      <title>Example Title 2</title>
      <description><![CDATA[Example description]]></description>
      <content:encoded><![CDATA[<p>Example content</p>]]></content:encoded>
      <link>https://example.com/articles/2</link>
      <guid isPermaLink="true">https://example.com/articles/2</guid>
      <pubDate>Sun, 27 Apr 2025 03:24:20 GMT</pubDate>
    </item>
  </channel>
</rss>'
`;
